controllers/k8ssandra/contact_points_service.go

```diff
@@ -0,0 +1,254 @@
+		Name:      dc.GetAllPodsServiceName(),
+	}
+	endpoints := &corev1.Endpoints{}
+	if err := remoteClient.Get(ctx, key, endpoints); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("Remote Endpoints not found", "key", key)
+			return nil, nil
+		}
+		logger.Error(err, "Failed to fetch remote Endpoints", "key", key)
+		return nil, err
+	}
+	if len(endpoints.Subsets) == 0 {
+		logger.Info("Remote Endpoints found but have no subsets", "key", key)
+		return nil, nil
+	}
+	logger.Info("Remote Endpoints found", "key", key)
+	return endpoints, nil
+}
+
+func contactPointsServiceKey(kc *api.K8ssandraCluster, dc *cassdcapi.CassandraDatacenter) client.ObjectKey {
+	return types.NamespacedName{
+		Namespace: kc.Namespace,
+		Name:      kc.SanitizedName() + "-" + dc.SanitizedName() + "-contact-points-service",
+	}
+}
+
+func (r *K8ssandraClusterReconciler) createContactPointsService(
+	ctx context.Context,
+	kc *api.K8ssandraCluster,
+	dc *cassdcapi.CassandraDatacenter,
+	remoteEndpoints *corev1.Endpoints,
+	logger logr.Logger,
+) error {
+	key := contactPointsServiceKey(kc, dc)
+	err := r.Client.Get(ctx, key, &corev1.Service{})
+	if err == nil {
+		// Service already exists, nothing to do.
+		// (note that we don't use a hash annotation here, because the contents are always the same)
+		return nil
+	}
+	if !errors.IsNotFound(err) {
+		logger.Error(err, "Failed to fetch Service", "key", key)
+		return err
+	}
+
+	// Else not found, create it
+	logger.Info("Creating Service", "key", key)
+	service, err := r.newContactPointsService(key, kc, dc, remoteEndpoints)
+	if err != nil {
+		logger.Error(err, "Failed to initialize Service", "key", key)
+		return err
+	}
+	if err = r.Client.Create(ctx, service); err != nil {
+		logger.Error(err, "Failed to create Service", "key", key)
+		return err
+	}
+	return nil
+}
+
+func (r *K8ssandraClusterReconciler) newContactPointsService(
+	key client.ObjectKey,
+	kc *api.K8ssandraCluster,
+	dc *cassdcapi.CassandraDatacenter,
+	remoteEndpoints *corev1.Endpoints,
+) (*corev1.Service, error) {
+	var ports []corev1.ServicePort
+	for _, remotePort := range remoteEndpoints.Subsets[0].Ports {
+		ports = append(ports, corev1.ServicePort{
+			Name:     remotePort.Name,
+			Port:     remotePort.Port,
+			Protocol: remotePort.Protocol,
+		})
+	}
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: key.Namespace,
+			Name:      key.Name,
+			Labels: map[string]string{
+				api.NameLabel:                      api.NameLabelValue,
+				api.PartOfLabel:                    api.PartOfLabelValue,
+				api.ComponentLabel:                 api.ComponentLabelValueCassandra,
+				api.K8ssandraClusterNameLabel:      kc.Name,
+				api.K8ssandraClusterNamespaceLabel: kc.Namespace,
+				api.DatacenterLabel:                dc.Name,
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Type:  corev1.ServiceTypeClusterIP,
+			Ports: ports,
+			// We don't provide a selector since the operator manages the Endpoints directly
```

Comment:
I'm not sure if this service should be headless (currently it is not), it seems convenient to use the service name the contact point if a component wants to connect to "any node" in the DC.

Reply:
Actually, I think the service _must_ be headless if we want to be able to resolve individual pod addresses.